airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt

```diff
@@ -47,9 +68,103 @@ abstract class AbstractStreamOperation<DestinationState : MinimumDestinationStat
         updatedDestinationState = destinationInitialStatus.destinationState.withSoftReset(false)
     }
 
-    companion object {
-        private const val NO_SUFFIX = ""
-        private const val TMP_OVERWRITE_TABLE_SUFFIX = "_airbyte_tmp"
+    private fun prepareStageForNormalSync(
+        stream: StreamConfig,
+        destinationInitialStatus: DestinationInitialStatus<DestinationState>
+    ): InitialRawTableStatus {
+        log.info {
+            "${stream.id.originalNamespace}.${stream.id.originalName}: non-truncate sync. Creating raw table if not exists."
+        }
+        storageOperation.prepareStage(stream.id, NO_SUFFIX)
+        if (destinationInitialStatus.initialTempRawTableStatus.rawTableExists) {
+            log.info {
+                "${stream.id.originalNamespace}.${stream.id.originalName}: non-truncate sync, but temp raw table exists. Transferring it to real raw table."
+            }
+            // There was a previous truncate refresh attempt, which failed, and left some
+            // records behind.
+            // Retrieve those records and put them in the real stage.
+            // This is necessary to avoid certain data loss scenarios.
+            // (specifically: a user initiates a truncate sync, which fails, but emits some records.
+            // It also emits a state message for "resumable" full refresh.
+            // The user then initiates an incremental sync, which runs using that state.
+            // In this case, we MUST retain the records from the truncate attempt.)
+            storageOperation.transferFromTempStage(stream.id, TMP_TABLE_SUFFIX)
+
+            // We need to combine the raw table statuses from the real and temp raw tables.
+            val hasUnprocessedRecords =
+                destinationInitialStatus.initialTempRawTableStatus.hasUnprocessedRecords ||
+                    destinationInitialStatus.initialRawTableStatus.hasUnprocessedRecords
+            // Pick the earlier min timestamp.
+            val maxProcessedTimestamp: Optional<Instant> =
+                destinationInitialStatus.initialRawTableStatus.maxProcessedTimestamp
```

Comment:
this seems very complex to just get the min timestamp (if I'm understanding this correcttly)

Reply:
yeah, I'd love a suggestion on how to do this better :/ 

(also, getting things prepped for monday release - any concern with merging this as-is + fixing in a followup?)

Reply:
going to just leave this as-is for now. there's probably something along these lines, though we probably don't _actually_ want to introduce [liftA2](https://hackage.haskell.org/package/base-4.20.0.1/docs/Prelude.html#v:liftA2) to kotlin :P

![image](https://github.com/airbytehq/airbyte/assets/5741425/68ad979d-afb2-4013-af38-d7955d69ee4d)